yt_dlp/utils/_utils.py

```diff
@@ -1364,22 +1364,34 @@ def datetime_add_months(dt_, months):
     return dt_.replace(year, month, day)
 
 
+def datetime_from_timestamp(timestamp):
+    # Working around out-of-range timestamp values (e.g. negative ones on Windows,
+    # see http://bugs.python.org/issue1646728)
+    # Using naive datetime here can break timestamp() in Windows
+    # Ref: https://github.com/yt-dlp/yt-dlp/issues/5185, https://github.com/python/cpython/issues/94414
+    # Also, dt.datetime.fromtimestamp breaks for negative timestamps
+    # Ref: https://github.com/yt-dlp/yt-dlp/issues/6706#issuecomment-1496842642
+    return (dt.datetime.fromtimestamp(0, dt.timezone.utc)
+            + dt.timedelta(seconds=timestamp))
+
+
 def datetime_round(dt_, precision='day'):
     """
     Round a datetime object's time to a specific precision
     """
     if precision == 'microsecond':
         return dt_
 
+    time_scale = 1000000
     unit_seconds = {
         'day': 86400,
         'hour': 3600,
         'minute': 60,
         'second': 1,
     }
     roundto = lambda x, n: ((x + n / 2) // n) * n
-    timestamp = roundto(calendar.timegm(dt_.timetuple()), unit_seconds[precision])
-    return dt.datetime.fromtimestamp(timestamp, dt.timezone.utc)
+    timestamp = roundto(calendar.timegm(dt_.timetuple()) * time_scale + dt_.microsecond, unit_seconds[precision] * time_scale) / time_scale
```

Comment:
isn't this cleaner?
```suggestion
    timestamp = roundto(calendar.timegm(dt_.timetuple()) + dt_.microsecond / 1E6, unit_seconds[precision])
```

Reply:
On second look, the `microseconds` don't do anything since that case is handled above already

Reply:
Hmm, I didn't get.

<br>

> `dt_.microsecond / 1E6`

`dt_.microsecond` is always less than `1E6`.

```console
> str(dt.datetime(1, 2, 3, 4, 5, 6, 7))
'0001-02-03 04:05:06.000007'

> str(dt.datetime(1, 2, 3, 4, 5, 6, 7777777))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: microsecond must be in 0..999999

```

<br>

> adding `microseconds` don't do anything since that case is handled above already

Not really, please see corresponding tests with `precision` set to 'second'.